xml_converter/src/xml_converter.cpp

```diff
@@ -130,50 +149,58 @@ void write_protobuf_file(string proto_filepath, map<string, Category>* marker_ca
         outfile.open(output_filepath, ios::out | ios::binary);
         output_message.CopyFrom(all_categories);
 
-        for (int i = 0; i < proto_pois.icon_size(); i++) {
-            if (proto_pois.icon(i).map_id() == map_id) {
-                vector<string> split_categories = split(proto_pois.icon(i).category().name(), ".");
-                auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
-                if (category == proto_categories.end()) {
-                    cout << "Unknown MarkerCategory " << proto_pois.icon(i).category().name() << endl;
-                }
-                else {
-                    // This loop adds the name of the category and all of it's parents
-                    string name;
-                    for (unsigned int j = 0; j < split_categories.size(); j++) {
-                        name += split_categories[j];
-                        category_includes.insert(name);
-                        name += ".";
+        for (const auto& parsed_poi : *parsed_pois) {
+            if (parsed_poi->classname() == "POI") {
+                Icon* icon = dynamic_cast<Icon*>(parsed_poi);
+                if (icon->map_id == map_id) {
+                    waypoint::Icon proto_icon = icon->as_protobuf();
+                    vector<string> split_categories = split(proto_icon.category().name(), ".");
+                    auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
+                    if (category == proto_categories.end()) {
+                        cout << "Unknown MarkerCategory " << proto_icon.category().name() << endl;
+                    }
+                    else {
+                        // This loop adds the name of the category and all of it's parents
+                        string name;
+                        for (unsigned int j = 0; j < split_categories.size(); j++) {
+                            name += split_categories[j];
+                            category_includes.insert(name);
+                            name += ".";
+                        }
                     }
+                    output_message.add_icon()->CopyFrom(proto_icon);
```

Comment:
Does this need to be a `CopyFrom()`? What is the memory ownership of the data?

Reply:
From my understanding of the documentation says that `CopyFrom()` and `MergeFrom()` are my best options for setting a repeated message. `Copyfrom()`: "The method clears the current message and then merges the specified message using MergeFrom." and `MergeFrom()` "This method merges the contents of the specified message into the current message. Singular fields that are set in the specified message overwrite the corresponding fields in the current message. Repeated fields are appended. Singular sub-messages and groups are recursively merged." 
There is no function such as `set_icon()` available. 

So in these instances, I think `MergeFrom()` is more appropriate since `add_icon()` returns a pointer to an empty message and thus there is no data to clear. 
The message itself is actually created in the `as_protobuf()` function. I have changed the code so that the output of this function goes directly into `out_message` instead of a separate variable.